rest-api/src/main/java/com/cloudbees/workflow/flownode/FlowNodeUtil.java

```diff
@@ -49,28 +71,113 @@
     private FlowNodeUtil() {
     }
 
-    public static long getNodeExecDuration(FlowNode node) {
-        List<FlowNode> childNodes = getChildNodes(node);
-        if (!childNodes.isEmpty()) {
-            long startTime = TimingAction.getStartTime(node);
-            long endTime = TimingAction.getStartTime(childNodes.get(childNodes.size() - 1));
+    public abstract static class CacheExtensionPoint implements ExtensionPoint {
```

Comment:
It _is_ simple and straightforward if you do it right. You just need one extension, with all the state.

```java
@Extension public final class FlowNodeCache {
    public static FlowNodeCache get() {
        return ExtensionList.lookup(FlowNodeCache.class).iterator().next();
    }
    private final Cache<String,Integer> exampleField = …;
    public int count(String key) {return exampleField.get(key);}
}
```

In the short term, callers can use `get()`:

```java
int count = FlowNodeCache.get().count("hello");
```

But if the caller is itself an extension, it is easier:

```java
@Extension public class Whatever extends RootAction {
    @Inject private FlowNodeCache cache;
    public void doStuff(@QueryParameter String stuff) {
        int count = cache.count(stuff);
        // …
    }
}
```

and when the caller is not an extension, you just thread the `FlowNodeCache` singleton instance through your method calls.

Unit tests need merely construct an empty cache object and pass them around.